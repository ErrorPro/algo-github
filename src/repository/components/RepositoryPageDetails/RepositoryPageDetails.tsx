import { Calendar, Eye, GitFork, Star } from "lucide-react";
import type { Repository } from "@/repository/types";

interface RepositoryPageDetailsProps {
  repository: Repository;
  onSetFavorite?: (id: number, isFavorite: boolean) => void;
}

export default function RepositoryPageDetails({
  repository,
  onSetFavorite,
}: RepositoryPageDetailsProps) {
  return (
    <main className="mx-auto max-w-4xl px-4 py-8 sm:px-6 lg:px-8">
      <div className="mb-8">
        <h1 className="text-4xl font-bold text-white md:text-5xl">
          {repository.name}
        </h1>
        <p className="mt-4 text-lg text-slate-300">{repository.description}</p>
      </div>

      <div className="mb-8 grid grid-cols-2 gap-4 sm:grid-cols-4">
        <div className="rounded-lg border border-slate-700 bg-slate-800 p-4">
          <div className="flex items-center gap-2">
            <Star className="h-5 w-5 text-yellow-400 fill-yellow-400" />
            <div>
              <p className="text-sm text-slate-400">Stars</p>
              <p className="text-2xl font-bold text-white">
                {repository.stars?.toLocaleString()}
              </p>
            </div>
          </div>
        </div>

        <div className="rounded-lg border border-slate-700 bg-slate-800 p-4">
          <div className="flex items-center gap-2">
            <GitFork className="h-5 w-5 text-blue-400" />
            <div>
              <p className="text-sm text-slate-400">Forks</p>
              <p className="text-2xl font-bold text-white">
                {repository.forks.toLocaleString()}
              </p>
            </div>
          </div>
        </div>

        <div className="rounded-lg border border-slate-700 bg-slate-800 p-4">
          <div className="flex items-center gap-2">
            <Eye className="h-5 w-5 text-green-400" />
            <div>
              <p className="text-sm text-slate-400">Watchers</p>
              <p className="text-2xl font-bold text-white">
                {repository.watchers.toLocaleString()}
              </p>
            </div>
          </div>
        </div>

        <div className="rounded-lg border border-slate-700 bg-slate-800 p-4">
          <div className="flex items-center gap-2">
            <Calendar className="h-5 w-5 text-purple-400" />
            <div>
              <p className="text-sm text-slate-400">Updated</p>
              <p className="text-lg font-bold text-white">
                {repository.updatedAt}
              </p>
            </div>
          </div>
        </div>
      </div>

      <div className="space-y-6">
        <div className="rounded-lg border border-slate-700 bg-slate-800 p-6">
          <h2 className="mb-4 text-xl font-semibold text-white">Language</h2>
          <div className="flex items-center gap-2">
            <div className="h-3 w-3 rounded-full bg-blue-500" />
            <span className="text-slate-300">{repository.language}</span>
          </div>
        </div>

        {repository.topics?.length ? (
          <div className="rounded-lg border border-slate-700 bg-slate-800 p-6">
            <h2 className="mb-4 text-xl font-semibold text-white">Topics</h2>
            <div className="flex flex-wrap gap-2">
              {repository.topics.map((topic) => (
                <span
                  key={topic}
                  className="rounded-full bg-slate-700 px-3 py-1 text-sm text-slate-300 hover:bg-slate-600 transition-colors cursor-pointer"
                >
                  {topic}
                </span>
              ))}
            </div>
          </div>
        ) : null}

        <div className="rounded-lg border border-slate-700 bg-slate-800 p-6">
          <div className="flex flex-col gap-3 sm:flex-row">
            <a
              href={repository.url}
              target="_blank"
              rel="noopener noreferrer"
              className="inline-block flex-1 rounded-lg bg-blue-600 px-6 py-3 text-center font-semibold text-white transition-colors hover:bg-blue-700"
            >
              View on GitHub
            </a>
            <button
              onClick={() =>
                onSetFavorite?.(repository.id, !repository.isFavorite)
              }
              className="inline-block flex-1 rounded-lg bg-slate-700 px-6 py-3 text-center font-semibold text-white transition-colors hover:bg-slate-600"
            >
              {repository.isFavorite
                ? "Remove from Favorites"
                : "Add to Favorites"}
            </button>
          </div>
        </div>
      </div>
    </main>
  );
}
