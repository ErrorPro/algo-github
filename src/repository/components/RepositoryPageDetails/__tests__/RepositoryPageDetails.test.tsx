import { render, screen, fireEvent } from "@testing-library/react";
import type { Repository } from "@/repository/types";

import RepositoryPageDetails from "../RepositoryPageDetails";

const mockRepository: Repository = {
  id: 1,
  name: "react-awesome-ui",
  description: "A beautiful and responsive React UI component library",
  stars: 2543,
  forks: 342,
  watchers: 128,
  language: "TypeScript",
  updatedAt: "2025-10-15",
  url: "https://github.com/example/react-awesome-ui",
  topics: ["react", "ui-components", "typescript"],
  isFavorite: false,
};

describe("RepositoryPageDetails", () => {
  it("should match shapshot", () => {
    const { container } = render(
      <RepositoryPageDetails repository={mockRepository} />
    );
    expect(container).toMatchSnapshot();
  });

  it("should render repository title", () => {
    render(<RepositoryPageDetails repository={mockRepository} />);
    expect(screen.getByText("react-awesome-ui")).toBeInTheDocument();
  });

  it("should render repository description", () => {
    render(<RepositoryPageDetails repository={mockRepository} />);
    expect(
      screen.getByText("A beautiful and responsive React UI component library")
    ).toBeInTheDocument();
  });

  it("should render stars count", () => {
    render(<RepositoryPageDetails repository={mockRepository} />);
    expect(screen.getByText("2,543")).toBeInTheDocument();
  });

  it("should render forks count", () => {
    render(<RepositoryPageDetails repository={mockRepository} />);
    expect(screen.getByText("342")).toBeInTheDocument();
  });

  it("should render watchers count", () => {
    render(<RepositoryPageDetails repository={mockRepository} />);
    expect(screen.getByText("128")).toBeInTheDocument();
  });

  it("should render updated date", () => {
    render(<RepositoryPageDetails repository={mockRepository} />);
    expect(screen.getByText("2025-10-15")).toBeInTheDocument();
  });

  it("should render language", () => {
    render(<RepositoryPageDetails repository={mockRepository} />);
    expect(screen.getByText("TypeScript")).toBeInTheDocument();
  });

  it("should render all topics", () => {
    render(<RepositoryPageDetails repository={mockRepository} />);
    expect(screen.getByText("react")).toBeInTheDocument();
    expect(screen.getByText("ui-components")).toBeInTheDocument();
    expect(screen.getByText("typescript")).toBeInTheDocument();
  });

  it("should not render topics section if empty", () => {
    const repoWithoutTopics = { ...mockRepository, topics: [] };
    render(<RepositoryPageDetails repository={repoWithoutTopics} />);
    expect(screen.queryByText("Topics")).not.toBeInTheDocument();
  });

  it('should display "Add to Favorites" button when not favorited', () => {
    render(<RepositoryPageDetails repository={mockRepository} />);
    expect(screen.getByText("Add to Favorites")).toBeInTheDocument();
  });

  it('should display "Remove from Favorites" button when favorited', () => {
    const favoritedRepo = { ...mockRepository, isFavorite: true };
    render(<RepositoryPageDetails repository={favoritedRepo} />);
    expect(screen.getByText("Remove from Favorites")).toBeInTheDocument();
  });

  it("should call onSetFavorite with correct parameters when favorite button is clicked", () => {
    const onSetFavorite = jest.fn();
    render(
      <RepositoryPageDetails
        repository={mockRepository}
        onSetFavorite={onSetFavorite}
      />
    );

    const favoriteButton = screen.getByRole("button", {
      name: /add to favorites/i,
    });
    fireEvent.click(favoriteButton);

    expect(onSetFavorite).toHaveBeenCalledWith(1, true);
  });

  it("should call onSetFavorite with false when removing from favorites", () => {
    const onSetFavorite = jest.fn();
    const favoritedRepo = { ...mockRepository, isFavorite: true };
    render(
      <RepositoryPageDetails
        repository={favoritedRepo}
        onSetFavorite={onSetFavorite}
      />
    );

    const removeButton = screen.getByRole("button", {
      name: /remove from favorites/i,
    });
    fireEvent.click(removeButton);

    expect(onSetFavorite).toHaveBeenCalledWith(1, false);
  });

  it("should handle missing onSetFavorite callback gracefully", () => {
    render(<RepositoryPageDetails repository={mockRepository} />);
    const favoriteButton = screen.getByRole("button", {
      name: /add to favorites/i,
    });

    expect(() => {
      fireEvent.click(favoriteButton);
    }).not.toThrow();
  });
});
