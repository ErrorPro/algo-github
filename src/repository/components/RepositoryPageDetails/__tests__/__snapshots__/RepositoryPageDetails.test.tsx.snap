// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositoryPageDetails should match shapshot 1`] = `
<div>
  <main
    class="mx-auto max-w-4xl px-4 py-8 sm:px-6 lg:px-8"
  >
    <div
      class="mb-8"
    >
      <h1
        class="text-4xl font-bold text-white md:text-5xl"
      >
        react-awesome-ui
      </h1>
      <p
        class="mt-4 text-lg text-slate-300"
      >
        A beautiful and responsive React UI component library
      </p>
    </div>
    <div
      class="mb-8 grid grid-cols-2 gap-4 sm:grid-cols-4"
    >
      <div
        class="rounded-lg border border-slate-700 bg-slate-800 p-4"
      >
        <div
          class="flex items-center gap-2"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-star h-5 w-5 text-yellow-400 fill-yellow-400"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M11.525 2.295a.53.53 0 0 1 .95 0l2.31 4.679a2.123 2.123 0 0 0 1.595 1.16l5.166.756a.53.53 0 0 1 .294.904l-3.736 3.638a2.123 2.123 0 0 0-.611 1.878l.882 5.14a.53.53 0 0 1-.771.56l-4.618-2.428a2.122 2.122 0 0 0-1.973 0L6.396 21.01a.53.53 0 0 1-.77-.56l.881-5.139a2.122 2.122 0 0 0-.611-1.879L2.16 9.795a.53.53 0 0 1 .294-.906l5.165-.755a2.122 2.122 0 0 0 1.597-1.16z"
            />
          </svg>
          <div>
            <p
              class="text-sm text-slate-400"
            >
              Stars
            </p>
            <p
              class="text-2xl font-bold text-white"
            >
              2,543
            </p>
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border border-slate-700 bg-slate-800 p-4"
      >
        <div
          class="flex items-center gap-2"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-git-fork h-5 w-5 text-blue-400"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="12"
              cy="18"
              r="3"
            />
            <circle
              cx="6"
              cy="6"
              r="3"
            />
            <circle
              cx="18"
              cy="6"
              r="3"
            />
            <path
              d="M18 9v2c0 .6-.4 1-1 1H7c-.6 0-1-.4-1-1V9"
            />
            <path
              d="M12 12v3"
            />
          </svg>
          <div>
            <p
              class="text-sm text-slate-400"
            >
              Forks
            </p>
            <p
              class="text-2xl font-bold text-white"
            >
              342
            </p>
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border border-slate-700 bg-slate-800 p-4"
      >
        <div
          class="flex items-center gap-2"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-eye h-5 w-5 text-green-400"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M2.062 12.348a1 1 0 0 1 0-.696 10.75 10.75 0 0 1 19.876 0 1 1 0 0 1 0 .696 10.75 10.75 0 0 1-19.876 0"
            />
            <circle
              cx="12"
              cy="12"
              r="3"
            />
          </svg>
          <div>
            <p
              class="text-sm text-slate-400"
            >
              Watchers
            </p>
            <p
              class="text-2xl font-bold text-white"
            >
              128
            </p>
          </div>
        </div>
      </div>
      <div
        class="rounded-lg border border-slate-700 bg-slate-800 p-4"
      >
        <div
          class="flex items-center gap-2"
        >
          <svg
            aria-hidden="true"
            class="lucide lucide-calendar h-5 w-5 text-purple-400"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 2v4"
            />
            <path
              d="M16 2v4"
            />
            <rect
              height="18"
              rx="2"
              width="18"
              x="3"
              y="4"
            />
            <path
              d="M3 10h18"
            />
          </svg>
          <div>
            <p
              class="text-sm text-slate-400"
            >
              Updated
            </p>
            <p
              class="text-lg font-bold text-white"
            >
              2025-10-15
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      class="space-y-6"
    >
      <div
        class="rounded-lg border border-slate-700 bg-slate-800 p-6"
      >
        <h2
          class="mb-4 text-xl font-semibold text-white"
        >
          Language
        </h2>
        <div
          class="flex items-center gap-2"
        >
          <div
            class="h-3 w-3 rounded-full bg-blue-500"
          />
          <span
            class="text-slate-300"
          >
            TypeScript
          </span>
        </div>
      </div>
      <div
        class="rounded-lg border border-slate-700 bg-slate-800 p-6"
      >
        <h2
          class="mb-4 text-xl font-semibold text-white"
        >
          Topics
        </h2>
        <div
          class="flex flex-wrap gap-2"
        >
          <span
            class="rounded-full bg-slate-700 px-3 py-1 text-sm text-slate-300 hover:bg-slate-600 transition-colors cursor-pointer"
          >
            react
          </span>
          <span
            class="rounded-full bg-slate-700 px-3 py-1 text-sm text-slate-300 hover:bg-slate-600 transition-colors cursor-pointer"
          >
            ui-components
          </span>
          <span
            class="rounded-full bg-slate-700 px-3 py-1 text-sm text-slate-300 hover:bg-slate-600 transition-colors cursor-pointer"
          >
            typescript
          </span>
        </div>
      </div>
      <div
        class="rounded-lg border border-slate-700 bg-slate-800 p-6"
      >
        <div
          class="flex flex-col gap-3 sm:flex-row"
        >
          <a
            class="inline-block flex-1 rounded-lg bg-blue-600 px-6 py-3 text-center font-semibold text-white transition-colors hover:bg-blue-700"
            href="https://github.com/example/react-awesome-ui"
            rel="noopener noreferrer"
            target="_blank"
          >
            View on GitHub
          </a>
          <button
            class="inline-block flex-1 rounded-lg bg-slate-700 px-6 py-3 text-center font-semibold text-white transition-colors hover:bg-slate-600"
          >
            Add to Favorites
          </button>
        </div>
      </div>
    </div>
  </main>
</div>
`;
